TGUI Say: Upgrades chat input with modern features

jlsnow301 opened the pull request:
<!-- Write **BELOW** The Headers and **ABOVE** The comments else it may not be viewable. -->
<!-- You can view Contributing.MD for a detailed description of the pull request process. -->

## About The Pull Request
- Creates a super streamlined tgui-say package that's independent of TGUI
- Extensive, separate logging courtesy of @Iamgoofball 
- Ports typing indicators influenced by @Kapu1178 
- Adds forced speech mechanics
- User prefs to disable typing indicators and tgui-say themes

<details>
<summary>Feature previews</summary>

window autoexpands on entry
![auto expand](https://user-images.githubusercontent.com/42397676/173295383-d2978988-c8f5-4a50-a179-cc5eadcf7a09.gif)

five message history: use arrows up/down
![chat history](https://user-images.githubusercontent.com/42397676/173777246-a3321469-5ab2-445a-abb2-66d5b7490678.gif)

light mode
![lightmode](https://user-images.githubusercontent.com/42397676/173777299-54821553-6c27-434c-b3ed-0c12a3cad0dc.png)

change channels by TAB/button OR typing department into me/radio
![channels](https://user-images.githubusercontent.com/42397676/173295417-85c38cc2-cfe1-4de5-aadc-ffcf597f56bf.gif)

quick radio hotkey "Y"
![radio hotkey](https://user-images.githubusercontent.com/42397676/173295481-e67879e2-165f-4145-87df-dfcbccce4650.gif)

thinking (window open) and typing indicators with inactivity detection
note: indicators can be turned off, do not show up for "secret" channels like OOC and binary
![typing indicators](https://user-images.githubusercontent.com/42397676/173295577-f3c6f2a6-1f90-4df7-b8fb-8db80f4a3586.gif)

forced speech: getting hurt *can* purge input
![forced speech](https://user-images.githubusercontent.com/42397676/173295720-d87838e2-ac98-4822-82df-3a47f0ee5b81.gif)

</details>
<!-- Describe The Pull Request. Please be sure every change is documented or this can delay review and even discourage maintainers from merging your PR! -->


## Why It's Good For The Game
Finally we shirk off the most tenacious part of this 1996 UI with a sleek and modern replacement that's CUSTOMIZABLE
<!-- Argue for the merits of your changes and how they benefit the game, especially if they are controversial and/or far reaching. If you can't actually explain WHY what you are doing will improve the game, then it probably isn't good for the game in the first place. -->

## Changelog

<!-- If your PR modifies aspects of the game that can be concretely observed by players or admins you should add a changelog. If your change does NOT meet this description, remove this section. Be sure to properly mark your PRs to prevent unnecessary GBP loss. You can read up on GBP and it's effects on PRs in the tgstation guides for contributors. Please note that maintainers freely reserve the right to remove and add tags should they deem it appropriate. You can attempt to finagle the system all you want, but it's best to shoot for clear communication right off the bat. -->


:cl: jlsnow301 KubeRoot stylemistake Iamgoofball Kapu1178
add: Adds the TGUI say modal to replace speech input boxes. Many features!
add: You can switch output channel within the TGUI say modal by clicking the channel button or keying TAB. 
add: Added 5-message chat history to the say window. Use arrow keys to scroll recent messages.
add: The say window auto-expands! It adds up to two extra lines and scrolls beyond that.
add: The common radio hotkey - "Y". No need to type ; in radio channel. Help maint!
add: Radio subchannel helpers - Typing subchannels like ":e " will show labels.
add: Getting hurt while speaking has a chance to force you to speak. You read that correc-AUGH
add: Typing indicators! Having an IC channel open will show thinking/typing icons activity.
add: Typing indicators are a preference. Go incognito if you want!
add: Themes for TGUI Say! There is currently a light mode. More to come.
refactor: Sorts other TGUI inputs into their own folder and cleans up their code. Remember, TGUI Inputs can be disabled in UI prefs if you're not a fan.
refactor: Refactors a lot of human defense code, report bugs if you encounter them.
refactor: Refactors admin centcom reports into typescript. You must send the report to save the text.
/:cl:

<!-- Both :cl:'s are required for the changelog to work! You can put your name to the right of the first :cl: if you want to overwrite your GitHub username as author ingame. -->
<!-- You can use multiple of the same prefix (they're only used for the icon ingame) and delete the unneeded ones. Despite some of the tags, changelogs should generally represent how a player might be affected by the changes rather than a summary of the PR's contents. -->

jlsnow301 committed:
Initial commit

Kapu1178 commented:
Oh boy replacing the single most important and already functional part of the game with an external dependancy that many players have issues with randomly breaking.

SoSMaster9000 commented:
Oh god oh fuck please god no oh god oh fuck

optimumtact commented:
> Oh boy replacing the single most important and already functional part of the game with an external dependancy that many players have issues with randomly breaking.

"many players"

MrsTonedOne reviewed (changes requested):
This **has** to be entirely client side to get my sign off.

There must be no delay on an nz <-> us connection from hitting `t` in hotkey mode, and being able to type something to say. ie, if you hit t and start typing what you want to say without looking at the screen.

it **is** possible to do this all client side, but this requires creating a client side window manager that can bind to that key press client side (possible, client side macrosets can target `.output` and use that to send a message to a browser javascript function) and handle presenting windows.

we have long needed more use of javascript for advance client side functionality, but nobody wants to spend the time to make it and instead wants to make shitty server side solutions to non issues.

the input dialog is *functional* i do not give a shit how ugly or old it looks, this is a **SECONDARY** concern to its ability to do the basic functional shit.

Kapu1178 commented:
> This **has** to be entirely client side to get my sign off.
> 
> There must be no delay on an nz <-> us connection from hitting `t` in hotkey mode, and being able to type something to say. ie, if you hit t and start typing what you want to say without looking at the screen.
> 
> it **is** possible to do this all client side, but this requires creating a client side window manager that can bind to that key press client side (possible, client side macrosets can target `.output` and use that to send a message to a browser javascript function) and handle presenting windows.
> 
> we have long needed more use of javascript for advance client side functionality, but nobody wants to spend the time to make it and instead wants to make shitty server side solutions to non issues.
> 
> the input dialog is _functional_ i do not give a shit how ugly or old it looks, this is a **SECONDARY** concern to its ability to do the basic functional shit.

I know deep inside TGUI's core is byond `winset()` calls. Beestation and soon Goonstation use a byond skin with verbs attached to all the functions to create a completely client side input. I'm sure there's a way to wrap this into a "pretty" TGUI bundle.

https://github.com/BeeStation/BeeStation-Hornet/pull/5417

jlsnow301 committed:
Entirely unpooled

jlsnow301 committed:
Inline

Mothblocks commented:
I've echoed my concerns on this before several times, and in several PRs.

**Pressing T __must__ instantly open up Say. There is no world in which I would accept a PR that does not do that.**

Jolly-66 commented:
Weren't we against this due to TGUI having the tendency to just, die sometimes?

jlsnow301 committed:
Breakpoint

stylemistake commented:
You used a normal tgui pipeline here instead of /datum/tgui_window, in almost every way similar to tgui_input, with external assets and with all the same problems, so...

You will have to scrap it and redo it completely.

Also see MSO's comment, but the main priority:

- Has to use /tgui/window
- **No external assets**, everything inlined.
- Has to be statically declared in skin.dmf, always in background, initialized like stat/chat.
- Opens and closes immediately on pressing T using Byond macros

jlsnow301 committed:
Rewrites tgui say and shuffles tgui input

jlsnow301 committed:
Linter

jlsnow301 committed:
Comment

jlsnow301 committed:
sleepy

jlsnow301 committed:
Tinkering

Mothblocks commented on tgui/packages/tgui-modal/index.js:
```diff
@@ -15,10 +14,15 @@ const setupApp = () => {
+    setupHotReloading();
+    module.hot.accept([
+      './components',
+    ], () => {
+      renderApp();
+    });
```
```suggestion
    ], renderApp);
```

Right?

jlsnow301 replied:
this is copy pasted directly over from main tgui package. maybe? this syntax is a little wonky

stylemistake replied:
If this is going to be inlined into html, then hot reloading is not needed at all.

jlsnow301 committed:
Update tgui/packages/tgui-modal/index.js

Co-authored-by: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

jlsnow301 committed:
nonfunctional

jlsnow301 committed:
Up and running

jlsnow301 committed:
Working

jlsnow301 committed:
fixes window

jlsnow301 committed:
Looking nice

jlsnow301 committed:
Before adding channels

jlsnow301 committed:
pre-bittling

jlsnow301 committed:
refactors tgui inputs

jlsnow301 committed:
gradient channels done

jlsnow301 committed:
Pre-biddling

jlsnow301 committed:
pre-props

jlsnow301 committed:
Working client modal

jlsnow301 committed:
Input refactor + This is working ha

jlsnow301 committed:
Input defactor

jlsnow301 committed:
Textarea onEnter

jlsnow301 committed:
Sorry formatter

jlsnow301 committed:
fixes tab

jlsnow301 committed:
configuring props

jlsnow301 committed:
Componentdidmount version

jlsnow301 committed:
Fixing props

jlsnow301 committed:
Merge remote-tracking branch 'upstream/master' into tgui-say

jlsnow301 committed:
stopping point

jlsnow301 committed:
Glorfened

jlsnow301 committed:
glorf is working

jlsnow301 committed:
glorf optimized + sanitized

jlsnow301 committed:
Adds force_say()

jlsnow301 committed:
Adding chat history

jlsnow301 committed:
fixes messaging + chat history

thx anturk.

Co-Authored-By: AnturK <AnturK@users.noreply.github.com>

jlsnow301 committed:
more testing + optimization

jlsnow301 committed:
cleanup, pre-chat bubble

jlsnow301 committed:
Typing indicators

jlsnow301 committed:
Fixes some bugs

Fikou commented:
can you make glorfing and indicators config?

Iamgoofball commented:
@Fikou why would we shut off indicators or glorf

Fikou commented on code/modules/mob/typing_indicator.dm:
```diff
@@ -0,0 +1,97 @@
+/mob/Logout()
+	remove_typing_indicator()
+	remove_thinking_indicator()
+	. = ..()
```
return ..() at end of procs

Fikou replied:
return ..() at end of procs

Fikou commented on code/modules/mob/typing_indicator.dm:
```diff
@@ -0,0 +1,97 @@
+/// Shows the thinking indicator - player has window open
+/mob/verb/start_thinking(channel as text)
+	set name = ".start_thinking"
+	set hidden = 1
```
can these not be TRUE?

Fikou commented on code/modules/mob/typing_indicator.dm:
```diff
@@ -0,0 +1,97 @@
+/// Shows the thinking indicator - player has window open
+/mob/verb/start_thinking(channel as text)
+	set name = ".start_thinking"
+	set hidden = 1
```
can these not be TRUE?

jlsnow301 replied:
Yes... I forgot to convert this over 😓

Fikou commented on code/modules/tgui_input/alert.dm:
```diff
@@ -0,0 +1,130 @@
+/datum/tgui_alert/Destroy(force, ...)
+	SStgui.close_uis(src)
+	QDEL_NULL(buttons)
+	. = ..()
```
return ..() at the end of procs

Fikou replied:
return ..() at the end of procs

Iamgoofball committed:
Logging + Typing Indicator Preference

jlsnow301 committed:
Merge pull request #4 from Iamgoofball/viva

Logging + Typing Indicator Preference

Iamgoofball committed:
whoops

Iamgoofball committed:
aaa

jlsnow301 committed:
Refactoring modal

jlsnow301 committed:
Credits + Reviews

Forgot to tag all of the people that made this PR possible. Thanks for your contributions

Co-Authored-By: Aleksej Komarov <stylemistake@gmail.com>
Co-Authored-By: KubeRoot <6917698+KubeRoot@users.noreply.github.com>
Co-Authored-By: Kapu1178 <75460809+Kapu1178@users.noreply.github.com>
Co-Authored-By: Iamgoofball <4081722+Iamgoofball@users.noreply.github.com>
Co-Authored-By: AnturK <AnturK@users.noreply.github.com>

Mannybrado commented:
hi

jlsnow301 committed:
Update tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/typing_indicator.tsx

jlsnow301 committed:
Merge pull request #5 from Iamgoofball/viva

whoops

jlsnow301 committed:
Some bulletproofing

jlsnow301 committed:
consolidation

jlsnow301 committed:
won't really need this

Kapu1178 commented on code/modules/client/client_procs.dm:
```diff
@@ -1036,18 +1036,25 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
 				if("Say")
-					// I'm so sorry for anyone that has to read this. This is strung together with trial & error.
-					var/say = "\".winset \\\"command=\\\".start_thinking say\\\";tgui_modal.is-visible=true;\\\"\""
+					var/say = create_window_message(SAY_CHAN)
```
theres no way this keeps this proc client side, right? Im pretty sure client verbs still query the server

jlsnow301 replied:
This should be just at keybinding assignment, the actual keystroke is using the string for .output (client side)
With this it's going key -> JS -> typing indicator
Previous was key -> indicator -> JS
I'll check in a minute

jlsnow301 replied:
and yes, this only runs on keybind init.

stylemistake replied:
If you can help it, set these up on the client (js), because if that say window stops working, people won't be able to OOC at all. Treat this as a fallback solution.

jlsnow301 committed:
Cleanup

jlsnow301 committed:
fixes focus

jlsnow301 committed:
Cleanups

jlsnow301 committed:
Oopsie

stylemistake commented on code/modules/client/client_procs.dm:
```diff
@@ -1031,12 +1035,25 @@ GLOBAL_LIST_INIT(blacklisted_builds, list(
+					var/ooc = create_window_message(OOC_CHAN)
+					winset(src, "default-[REF(key)]", "parent=default;name=[key];command=[ooc]")
+
+/** Creates a JSON encoded message to open TGUI modals properly */
+/client/proc/create_window_message(channel)
+	var/message = TGUI_CREATE_MESSAGE("open", list(
+		channel = channel,
+	))
+	return "\".output tgui_modal.browser:update [message]\""
```
This proc name is inaccurate in regards to what it does.

jlsnow301 replied:
create_output_string ?

stylemistake replied:
tgui_modal_create_open_command

stylemistake commented on tgui/packages/tgui/components/TextArea.js:
```diff
@@ -104,24 +130,19 @@ export class TextArea extends Component {
-          input.select();
-        }
-      }, 1);
+      Byond.subscribeTo('open', () => {
+        this.getFocus();
+      });
+      this.getFocus();
```
Please revert this, first of all subscribeTo is going to be a huge memory leak in a commonly used component like this (because unsubscribe is not implemented), second of all tgui host does not implement the "open" message, so this is completely specific to your tgui-modal.

Instead implement focusing logic yourself in tgui-modal. Implement an `innerRef` property that you could pass to this component (would initialize `this.textareaRef` if innerRef property is defined), this would allow getting a reference to textarea and doing whatever you want to it tgui-modal.

stylemistake commented on code/modules/tgui_panel/external.dm:
```diff
@@ -4,6 +4,7 @@
  */
 
 /client/var/datum/tgui_panel/tgui_panel
+/client/var/datum/tgui_modal/tgui_modal
```
Create a tgui_modal module and try to keep as many procs and vars contained in it as possible, that includes all client procs and vars.

stylemistake commented on code/modules/tgui/tgui.dm:
```diff
@@ -47,12 +47,10 @@
  * return datum/tgui The requested UI.
  */
-/datum/tgui/New(mob/user, datum/src_object, interface, title, ui_x, ui_y)
+/datum/tgui/New(mob/user, datum/src_object, interface, title)
```
This seems very out of scope, because you didn't completely remove all associated functionality. Revert.

stylemistake reviewed (changes requested):
Some feedback given, also your code is all over the place, please restructure and modularize it

blackdav123 commented:
bring back the cancel button or x button the old box had

jlsnow301 commented:
> bring back the cancel button or x button the old box had

this is the escape button

Astrogem2 commented:
i hate it i hate it i hate it i hate it i hate it, i want to be able to have old box.

jlsnow301 committed:
Review tending

CoffeeDragon16 commented:
Hello! Love the PR. Tgui says is a big move forward, and you're doing a good job. I think color coding the say type doesn't work well, though. I'm fairly colorblind, and I can't tell the difference between the OOC and radio color. I'd imagine other players are having problems like that. Maybe get rid of the > and just always keep the letter there, instead of on mouseover?

jlsnow301 committed:
Moves focus

good idea thx

Co-Authored-By: Aleksej Komarov <stylemistake@gmail.com>

jlsnow301 commented:
> Hello! Love the PR. Tgui says is a big move forward, and you're doing a good job. I think color coding the say type doesn't work well, though. I'm fairly colorblind, and I can't tell the difference between the OOC and radio color. I'd imagine other players are having problems like that. Maybe get rid of the > and just always keep the letter there, instead of on mouseover?

Sure! Accessibility is a top concern

jlsnow301 committed:
compacted window and fixed terminal issues

Mothblocks commented on code/modules/tgui_input/modal/speech.dm:
```diff
@@ -0,0 +1,105 @@
+	/// No OOC leaks
+	if(payload["channel"] == OOC_CHAN || payload["channel"] == ME_CHAN)
+		return pick(hurt_phrases)
+	/// Sanitizes radio prefixes so users can't game the system (mostly)
```
I said this in Discord earlier but I really don't think glorfing should remove radio prefixes, it will just incentivize more ";HELP" and punish people who can't type very fast

jlsnow301 replied:
it is done

jlsnow301 committed:
"better" timers for typing indicators

stylemistake commented:
> And I can't tell the difference between the OOC and radio color.

I think animated borders also hurt accessibility, I'd recommend sticking to my design.

jlsnow301 committed:
changes skin

jlsnow301 committed:
Changes glorf handling

jlsnow301 commented:
>I'd recommend sticking to my design.

Saw this after my most recent commit but it is closer to this

jlsnow301 committed:
Style fixes

jlsnow301 committed:
Misc fixes

jlsnow301 committed:
Some error cases

jlsnow301 committed:
Better button

jlsnow301 committed:
Send it

jlsnow301 committed:
Oop

Jamarkus127 commented:
This is nice, but its now not possible to queue what you were saying, or keep specific me verbs queued up in a corner. yeah, we have a five chat history, but that isnt a lot especially if you are speaking and emoting alot. it also dosnt help if I want to queue singing a song, or queue a list that I want to say one at a time. typing indicators are great though. I really like them.

Another issue is the wrapping text. In theory its nice. but now that there is an entire layer system to text, I want to Press the up arrow to get to that top layer of text. that erases my entire message, and I cant recover it by pressing down again. thats a big issue.

jlsnow301 commented:
> This is nice, but its now not possible to queue what you were saying, or keep specific me verbs queued up in a corner. 

there is only one window and it never technically goes away. please read earlier comments to see why this was the design.

for arrow keys - i can always change the key that does this but it would make less sense when you can also click where you want, mid sentence, very similar to how you'd do it on a phone

jlsnow301 committed:
Removes thinking indies from me/ooc

jlsnow301 committed:
adds window dragging

jlsnow301 committed:
thanks I hate it

jlsnow301 committed:
sticky window position

Pickle-Coding commented:
There's a bit of lag to opening the tgui chat window compared to normal, and this has made me accidentally miss the first 1-3 characters in whatever I am going to say quite a lot of times, which is annoying.

jlsnow301 commented:
> Also weird that typing indicators show when uou are typing in OOC.

this was changed

JohnFulpWillard commented on code/modules/tgui_input/modal/modal.dm:
```diff
@@ -0,0 +1,135 @@
+ * payload - A list containing the channel the window was opened in.
+ */
+/datum/tgui_modal/proc/open(payload)
+	if(!client?.mob)
+		CRASH(NULL_CLIENTMOB)
```
Here (and a few other places), don't we already expect a client to be using this (since it's TGUI)? Otherwise, can't we just check for !client and early return, instead of using ? everywhere

jlsnow301 replied:
very possible, i'm just being incredibly extra with regards to client

TheSilverNuke commented:
Speech bubble location should be changed. At the moment they're hidden behind health HUD icons, so you can't actually see if someone is talking due to overlap.

jlsnow301 commented:
i cannot ping pong the size and styling. the first version of this was larger and easy to read but received tons of negative feedback on how large it was. i can add options to set it to user prefs at a later date.

the issue i'm looking at currently is just typing indies being covered by huds

TheSilverNuke commented:
> i cannot ping pong the size and styling. the first version of this was larger and easy to read but received tons of negative feedback on how large it was. i can add options to set it to user prefs at a later date.
> 
> the issue i'm looking at currently is just typing indies being covered by huds

If people are making you constantly tweak the box, why not just create some presets? I'm not sure how difficult this would be, but if you were you could make it such that there were three options between small, medium, and large.

jlsnow301 committed:
resets winsize on history view

MrsTonedOne reviewed (approved):
Passes the client side lag check.

jlsnow301 committed:
Typing timer adjustment

cacogen commented:
couple of things i noticed playing the testmerge

1. if you open the chat window with t, it shows a typing indicator, but not if you use m or o (i assume this is deliberate), but if you tab through to me or ooc after opening the chat window with tab, the typing indicator doesn't go away
2. when you're hit, the window doesn't close and auto submit the text (the GLORF feature we used to have), which was both an interesting feature and a way to avoid people having their keystrokes captured by the text window when they're attacked and trying to evade it

jlsnow301 commented:
> 1. the typing indicator doesn't go away

typing indicators last serverside for about 6 seconds (give or take lag). so an ooc or me window will not tell it to delete, it just won't refresh the 6 second typing indicator.

> when you're hit, the window doesn't close and auto submit the text

The auto submit feature is working, consider that you must hit people with enough force to do it. They must be typing into an IC channel and it does not close the window, only clears their input. This was decided so players wouldnt continue keying and accidentally press a hotkey

jlsnow301 committed:
Review tending

jlsnow301 committed:
Cleanup

jlsnow301 committed:
Linter really

GoldenAlpharex commented:
> for arrow keys - i can always change the key that does this but it would make less sense than arrows. you can also click where you want, mid sentence, very similar to how you'd do it on a phone

What I'd suggest doing for this is probably something like only allowing it to happen once your typing bar has reached the top of the window, that you let go of the up arrow, and then pressed the up arrow again. It might not sound like much, but it can end up being a big annoyance, especially when you're proof-reading through a big emote you wrote. We shouldn't force users to use their mouse when they're using widespread keyboard features and stuff that was already implemented into the standard BYOND input boxes.

GoldenAlpharex commented on code/__DEFINES/speech_channels.dm:
```diff
@@ -0,0 +1,5 @@
+// Used to direct channels to speak into.
+#define SAY_CHAN "Say"
+#define RADIO_CHAN "Radio"
+#define ME_CHAN "Me"
+#define OOC_CHAN "OOC"
```
Just out of curiosity, why not make them `X_CHANNEL`? Are those three letter gained really making much of a difference to the decrease in readability, as little as it might arguably be?

GoldenAlpharex reviewed (changes requested):
On top of all of this, I also think that having the option for certain modals to have a different amount of lines for their box could be interesting, in cases where we'd, say, want to have more lines for a `me`, to encourage people being more verbose on the more MRP servers (and on higher roleplay downstreams, too).

Oh, and if you end up making the max number of lines a pref, I'd suggest only sending the info on initial pref load, and when the preference gets changed, to reduce the latency and payload tied to adding more parameters that are almost always going to remain constant for a particular user during their entire round.

Iamgoofball commented on tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/typing_indicator.tsx:
```diff
@@ -0,0 +1,8 @@
+import { CheckboxInput, FeatureToggle } from "../base";
+
+export const typingindicator: FeatureToggle = {
+  name: "Enable typing indicators for self",
+  category: "GAMEPLAY",
+  description: "Enable typing indicators that show you're typing a message.",
+  component: CheckboxInput,
+};
```
yes

Iamgoofball commented on tgui/packages/tgui/interfaces/PreferencesMenu/preferences/features/game_preferences/typing_indicator.tsx:
```diff
@@ -0,0 +1,8 @@
+import { CheckboxInput, FeatureToggle } from "../base";
+
+export const typingindicator: FeatureToggle = {
```
if you update this, update the key in code as well

Iamgoofball commented on code/modules/tgui_input/modal/modal.dm:
```diff
@@ -0,0 +1,135 @@
+	if(client.typing_indicators)
+		log_speech_indicators("[key_name(client)] stopped typing IC at [loc_name(client.mob)], indicators enabled.")
+	else
+		log_speech_indicators("[key_name(client)] stopped typing IC at [loc_name(client.mob)], indicators DISABLED.")
```
it's the non-standard, and thus should stand out

jlsnow301 commented on tgui/packages/tgui/components/TextArea.js:
```diff
@@ -54,19 +54,48 @@ export class TextArea extends Component {
     };
     this.handleKeyDown = e => {
       const { editing } = this.state;
-      const { onKeyDown } = this.props;
+      const { onChange, onInput, onEnter, onKeyDown } = this.props;
+      if (e.keyCode === KEY_ENTER) {
+        this.setEditing(false);
+        if (onChange) {
+          onChange(e, e.target.value);
+        }
+        if (onInput) {
+          onInput(e, e.target.value);
+        }
+        if (onEnter) {
+          onEnter(e, e.target.value);
+        }
+        if (this.props.selfClear) {
+          e.target.value = '';
+        } else {
+          e.target.blur();
+        }
+        return;
+      }
```
word for word entered over from input.js. There is a LOT I want to just completely refactor / rewrite (including typescript) and I feel that is out of scope for this pr

jlsnow301 commented on code/modules/tgui_input/modal/speech.dm:
```diff
@@ -0,0 +1,132 @@
+	switch(channel)
+		if(RADIO_CHAN)
+			entry = remove_prefixes(entry)
+			if(entry)
+				entry = ";" + entry
+				client.mob.say_verb(entry)
```
it does not work. radio is common

stylemistake commented on tgui/packages/tgui/components/TextArea.js:
```diff
@@ -104,7 +130,9 @@ export class TextArea extends Component {
       input.value = toInputValue(nextValue);
     }
-
+    if (this.props.setInnerRef) {
+      this.props.setInnerRef(this.textareaRef);
+    }
```
Not sure if I actually suggested you to do this, did you try doing `this.textareaRef = props.innerRef` and passing it a ref?

jlsnow301 commented on tgui/packages/tgui-modal/interfaces/TguiModal.js:
```diff
@@ -0,0 +1,287 @@
+    this.reset();
+    windowClose();
+  };
+  /** User presses escape, closes the window */
```
the comment blocks, which nearly no interface in the game even use, form the line break. this is such a hard bikeshed

jlsnow301 commented on code/modules/tgui_input/modal/speech.dm:
```diff
@@ -0,0 +1,132 @@
+		delegate_speech(payload["entry"], payload["channel"])
+		return TRUE
+	if(type == "force")
+		var/target_chan = payload["channel"]
```
it's also literally 3 letters

jlsnow301 committed:
typescript and reviews

jlsnow301 committed:
cleanup

jlsnow301 committed:
random ci fail

jlsnow301 committed:
Checks

TrashDoxx commented:
Good changes, btw👍🏻

MrStonedOne commented:
>by design i do NOT want to track client side input channel (defeats REST design) every time they switch channels. 

Toobad, you need to fix the bug or remove typing/thinking indicators all together

TrashDoxx commented:
Based.

Wallemations commented:
Currently, Medhuds completely cover the speech bubble, making it so ghosts/doctors can't really tell when a person's speaking. Not sure if there's any sort of easy fix for that, but just wanted to make sure it was down in text that this is a bit of an issue.

jlsnow301 committed:
sigh

jlsnow301 committed:
no need

jlsnow301 committed:
better not leave this in

jlsnow301 committed:
cooldowns

itseasytosee commented:
HOLY SHIT MAN THIS IS DEVINE MANA FLOWING IN MY VIENS AHHHHHH

jlsnow301 committed:
Fixes

jlsnow301 committed:
touches up icons

jlsnow301 commented:
> Currently, Medhuds completely cover the speech bubble, making it so ghosts/doctors can't really tell when a person's speaking. Not sure if there's any sort of easy fix for that, but just wanted to make sure it was down in text that this is a bit of an issue.

this is fixed

jlsnow301 committed:
More responsive?

Dorsisdwarf commented:
As somone who has to reload the browser UI to make TGUI work pretty much every third or fourth round and frequently has tgui popups slow, become unresponsive, or not load, I am extremely worried about this being the project's chosen method of removing the always-functional never-failing instant-responsive speech bar.

Iamgoofball commented:
Why don't you go try the testmerge and see how it works for you?

jlsnow301 commented:
> As somone who has to reload the browser UI to make TGUI work pretty much every third or fourth round and frequently has tgui popups slow, become unresponsive, or not load, I am extremely worried about this being the project's chosen method of removing the always-functional never-failing instant-responsive speech bar.

fourth bullet point

jlsnow301 committed:
removed abductor indicators

jlsnow301 committed:
fixes an issue with command reports

Co-Authored-By: DomitiusKnack <56321744+DomitiusKnack@users.noreply.github.com>

jlsnow301 committed:
hanako

jlsnow301 committed:
Department channels

thx mothblocks for the idea

Co-Authored-By: Mothblocks <35135081+Mothblocks@users.noreply.github.com>

jlsnow301 committed:
kimiko

jlsnow301 committed:
ok rename

jlsnow301 committed:
panam

Mothblocks commented:
im not sure i get what REST has to do with this?

jlsnow301 commented:
> im not sure i get what REST has to do with this?

this was fixed

Simplehorror commented:
you don't need this for typing indicators

jlsnow301 committed:
scss rocks

Justice12354 commented:
In Round 184748, on Terry, whenever I pressed Enter after writing anything on the TGUI, nothing would happen. Using the say verb on the command bar worked perfectly, though. Commit a72537a2

jlsnow301 committed:
Massive refactor

jlsnow301 committed:
Organization

jlsnow301 commented:
> In Round 184748, on Terry, whenever I pressed Enter after writing anything on the TGUI, nothing would happen. 

Thank you for the report. Do you have any other info on the state of the window? Did it work then stop, or did any other feature work (ie: tab keys, etc)

I will perform some more testing to see if I can find a cause.

jlsnow301 committed:
Eslint

jlsnow301 committed:
it's about sending a message

Justice12354 commented:
> > In Round 184748, on Terry, whenever I pressed Enter after writing anything on the TGUI, nothing would happen.
> 
> Thank you for the report. Do you have any other info on the state of the window? Did it work then stop, or did any other feature work (ie: tab keys, etc)
> 
> I will perform some more testing to see if I can find a cause.

The window looked fine. I only noticed this a few minutes after the round started, so I think it may've been caused by something in the round itself (or not, who knows?). Whenever I tried to use the channel prefixes, they worked normally (for instance, when I typed ":m", the "Med" symbol appeared). I didn't pay attention to anything else, since I spent the rest of the round working around the command bar. 

Someone else complained about this too in #help (on Discord) and they claimed that it fixed itself after reconnecting. According to them, the "Refresh TGUI" verb in the OOC Panel opened a white window, which they couldn't close.

FlamingCheese commented:
The asay/dsay prompts still use the old input box and for me the current font is harder to read than other fonts, a toggle to switch to the old window (like the tgui list/confirmation prompts/text field prompts you added) or customization options would be very welcome

jlsnow301 commented:
> The asay/dsay prompts still use the old input box and for me the current font is harder to read than other fonts, a toggle to switch to the old window (like the tgui list/confirmation prompts/text field prompts you added) or customization options would be very welcome

this is now fixed
turn off tgui inputs in prefs

jlsnow301 committed:
Refactor + Modal disable

NoxVS commented:
Had a round, 184804, where two players had issues with the chat box. They could open it up with a hotkey, start talking in it, but sending it would simply close the textbox without any message sent. It wouldn't send a message either ingame or in logs. Both players had this affecting different types of chat - One could talk in OOC but not IC, the other could talk in IC but not in OOC or in emotes. Both were solved by restarting BYOND.

Wallemations commented:
Would it be possible to add a speech bubble for /me text as well? I understand if this would be too annoying, however it would be helpful if you're trying to type something out for that.
I've seen other servers use a rectangle instead of a speech bubble so you can tell those apart, maybe that could work?

NoxVS commented:
There are things, such as ghost role prompts, that can take away focus from the box you are typing in. This happened before this PR as well but because you can only have one box open at a time with this PR, pressing any of the hotkeys that open a chat box will close the old one and delete whatever you are writing.

jlsnow301 committed:
larger text and permabutton

Co-Authored-By: LemonInTheDark <58055496+LemonInTheDark@users.noreply.github.com>

jlsnow301 commented:
> Would it be possible to add a speech bubble for /me text as well? I understand if this would be too annoying, however it would be helpful if you're trying to type something out for that. I've seen other servers use a rectangle instead of a speech bubble so you can tell those apart, maybe that could work?

thinking/typing indies will be enabled for me chat now
custom bubbles will be coming eventually/future patch

jlsnow301 committed:
typing indicators for /me

Co-Authored-By: Wallem <66052067+Wallemations@users.noreply.github.com>

jlsnow301 committed:
micro op

jlsnow301 committed:
styling

jlsnow301 committed:
bug fix

jlsnow301 committed:
more bug fixes

jlsnow301 committed:
client mob

jlsnow301 committed:
idk

jlsnow301 committed:
111

LemonInTheDark commented:
> > The asay/dsay prompts still use the old input box and for me the current font is harder to read than other fonts, a toggle to switch to the old window (like the tgui list/confirmation prompts/text field prompts you added) or customization options would be very welcome
> 
> this is now fixed turn off tgui inputs in prefs

Do things like glorf or the indicator work with the pref off

jlsnow301 commented:
> Do things like glorf or the indicator work with the pref off

No, it just gives them the old window (which was still accessible via verbs anywho)

jlsnow301 committed:
cleanup

jlsnow301 committed:
no longer clears data on pressing hotkey twice

jlsnow301 committed:
revert

LemonInTheDark commented:
> > Do things like glorf or the indicator work with the pref off
> 
> No, it just gives them the old window (which was still accessible via verbs anywho)

That's a shame, having ways to bypass the new features seems silly

jlsnow301 committed:
say less

jlsnow301 commented:
light mode now available in user prefs

jlsnow301 committed:
adds light mode

jlsnow301 committed:
oops

jlsnow301 committed:
random ci fail

jlsnow301 committed:
formatter :(

jlsnow301 committed:
prettier when

LukasBeedellCodestuff commented:
This isn't just a re-skin of the chat window and has HUGE balance applications. When you're hit the textbox closes and your character says "ergh, argh ect" meaning you can't call for help over coms. This is huge and I don't think should exist at all. The fact  that the whole text box closes is moronic and a huge buff for antags as now there is 0 way to call for help when you're being hit.

jlsnow301 commented:
> This isn't just a re-skin of the chat window and has HUGE balance applications. When you're hit the textbox closes and your character says "ergh, argh ect" meaning you can't call for help over coms.

This is in the changelog and pr desc. Forced speech will not ever be removed from this pr for the following reasons:
- It's hilarious

LukasBeedellCodestuff commented:
> 

I saw that it was written there but can you justify the balance implications of having people unable to speak whilst they're being hit?

jlsnow301 commented:
> I saw that it was written there but can you justify the balance implications of having people unable to speak whilst they're being hit?

Yes: it's funny
My only rule is not taking fun from the game

jlsnow301 committed:
debug artifact

Iamgoofball commented:
@LukasBeedellCodestuff It was a long time feature we had that was only removed because hotkey mode became prevalent, and it's being re-added now that the technical hurdle has been cleared.

jlsnow301 commented:
> I think it's a good idea to add a cooldown on glorfing later so that it's primarily for ambushes, and not sustained attacks

as an easier alternative: there is currently a small debounce timer on it that prevents it from firing too often, which can very easily be extended

Simplehorror commented:
> This isn't just a re-skin of the chat window and has HUGE balance applications. When you're hit the textbox closes and your character says "ergh, argh ect" meaning you can't call for help over coms. This is huge and I don't think should exist at all. The fact that the whole text box closes is moronic and a huge buff for antags as now there is 0 way to call for help when you're being hit.

use old gui

jlsnow301 commented:
> I hate to admit it, but this is a rare TG W. Looks amazing too. Good job.

Really means a lot from one of my role model coders, thank you Bacon
